src/gui/sdlmain.cpp

```diff
@@ -3829,15 +3819,7 @@ bool GFX_Events()
 #endif
 
 	SDL_Event event;
-#if defined (REDUCE_JOYSTICK_POLLING)
-	static auto last_check_joystick = GetTicks();
-	auto current_check_joystick = GetTicks();
-	if (GetTicksDiff(current_check_joystick, last_check_joystick) > 20) {
-		last_check_joystick = current_check_joystick;
-		if (MAPPER_IsUsingJoysticks()) SDL_JoystickUpdate();
-		MAPPER_UpdateJoysticks();
-	}
-#endif
+	MAPPER_UpdateJoysticks();
```

Comment:
Should we be checking `MAPPER_IsUsingJoysticks()` before calling this?  That's the way it was before.  Or does that break hotplug?

Also, I see we're not calling `SDL_JoystickUpdate()` anymore.  SDL says that function gets called automatically in the event loop if we have joystick events enabled (which we now do).

My concern is that we're now polling the joysticks _before_ the event loop runs.  So we're maybe getting slightly outdated input here.  I'm not sure this is actually problematic in practice but should this call be moved to the end of the function?

Reply:
Gonna look into this.

Reply:
> Should we be checking MAPPER_IsUsingJoysticks() before calling this? That's the way it was before. 

No it wasn't, because that `if` statement had no braces and only covered `SDL_JoystickUpdate()`.

You might wonder, well shouldn't we at least guard `MAPPER_UpdateJoysticks()`? No need, because when `mapper.sticks.num == 0`, it is also the case that `mapper.sticks.num_groups == 0`. More specifically, `mapper.sticks.num_groups <= mapper.sticks.num`. See `QueryJoysticks` and `CreateBindGroups`, taken together with `ClearSticks` and their relative call order.

And this is also enforced by the assertion that is the first line of the loop of `Mapper_UpdateJoysticks`():
```
bool MAPPER_IsUsingJoysticks() {
	return (mapper.sticks.num > 0);
}
void MAPPER_UpdateJoysticks() {
	for (Bitu i=0; i<mapper.sticks.num_groups; i++) {
		assert(mapper.sticks.stick[i]);
		mapper.sticks.stick[i]->UpdateJoystick();
	}
}
```




> My concern is that we're now polling the joysticks before the event loop runs. So we're maybe getting slightly outdated input here. I'm not sure this is actually problematic in practice but should this call be moved to the end of the function?

What happens between this line and the end of the function that updates the reported joystick state?

Reply:
Looking at `CStickBindGroup.CheckEvent`, it's not clear to me that we need to call `Mapper_UpdateJoysticks()` at all. That function grabs joystick updates from the events. Why would we need anything else?
```
	bool CheckEvent(SDL_Event * event) override {
		SDL_JoyAxisEvent * jaxis = nullptr;
		SDL_JoyButtonEvent *jbutton = nullptr;

		switch(event->type) {
			case SDL_JOYAXISMOTION:
				jaxis = &event->jaxis;
				if(jaxis->which == stick_id) {
					if(jaxis->axis == 0)
						JOYSTICK_Move_X(emustick, jaxis->value);
					else if (jaxis->axis == 1)
						JOYSTICK_Move_Y(emustick, jaxis->value);
				}
				break;
			case SDL_JOYBUTTONDOWN:
			case SDL_JOYBUTTONUP:
				jbutton = &event->jbutton;
			        if (jbutton->which != stick_id) {
				        break;
			        }
			        const bool state = jbutton->type == SDL_JOYBUTTONDOWN;
			        const auto but = check_cast<uint8_t>(
			                jbutton->button % emulated_buttons);
			        JOYSTICK_Button(emustick, but, state);
			        break;
		        }
		        return false;
	}
```

Gonna test this out without the function call.

Reply:
Yeah never mind, definitely need that call for joystick to work. Still curious about how to know when it should be called.

Reply:
The mapper isn't using SDL events to update joystick state.  It ends up calling `ActivateJoystickBoundEvents()` which calls `SDL_JoystickGetButton()` and `SDL_JoystickGetAxis()` to get the current state.  So it's essentially polling the joystick every time this function gets called rather than event handling.

>  What happens between this line and the end of the function that updates the reported joystick state?

`SDL_PollEvents()` happens.  I took a quick look at SDL source code to confirm what happens internally.  It doesn't actually poll the hardware when we make those SDL_Get* function calls.  It just reads from some internal state variables.

`SDL_PollEvents()` will make a call to `SDL_UpdateJoysticks()`which updates internal SDL joystick state variables.  It may either poll the hardware directly or use events in the queue (probably differs based on platform).

Bottom line, there may be a slight increase in input latency because we're updating our joystick state before SDL has ran its update.  Not sure if it's enough to be problematic in practice but worth mentioning.

Reply:
You're correct about `MAPPER_IsUsingJoysticks()` not being needed though.  I really dislike single line if statements for exactly that reason :laughing: 

Looks like `MAPPER_UpdateJoysticks()` is a no-op if there's no joysticks connected anyway so that part is fine as-is.